### PR TITLE
Remove function queues from k8s watchers

### PR DESCRIPTION
**Pull request by aanm**
k8s watcher library already provides a queue for each watcher so we
don't need our function queue implementation on top of it.

This means we will process all events upon initialization which can slow
down the initialization of the Cilium API. Some benchmarks were 
performed with the simulation of 50k pods and 2k nodes to see and Cilium took around 46 
seconds to initialize. Of course this highly depends in the network and
the compute power available to Cilium. This benchmarks were executed on
a VM with 2 CPUs and a kube-apiserver available in the same machine 
where the VM was executed.

**First 10 seconds of a pprof trace capture for a cilium-agent initialization with 10k simulated pods and 1k simulated nodes simulated. (top is master, bottom is this branch)** (the number of GC calls were smaller and took less amount of time):

**mem stats of a cilium-agent with 50k simulated pods and 2k simulated nodes. (top is master, bottom is this branch)** (the `total-alloc` decreased from `3GB` to `2.62GB`, sys decreased from `1.52GB` to `1.25GB`:
```
alloc: 706.19MB (740493104 bytes)
total-alloc: 3.03GB (3255584776 bytes)
sys: 1.52GB (1629220187 bytes)
lookups: 0
mallocs: 16221287
frees: 10367753
heap-alloc: 706.19MB (740493104 bytes)
heap-sys: 1.43GB (1535016960 bytes)
heap-idle: 638.75MB (669777920 bytes)
heap-in-use: 825.16MB (865239040 bytes)
heap-released: 39.98MB (41918464 bytes)
heap-objects: 5853534
stack-in-use: 8.09MB (8486912 bytes)
stack-sys: 8.09MB (8486912 bytes)
stack-mspan-inuse: 11.61MB (12173632 bytes)
stack-mspan-sys: 19.52MB (20463616 bytes)
stack-mcache-inuse: 3.39KB (3472 bytes)
stack-mcache-sys: 16.00KB (16384 bytes)
other-sys: 3.04MB (3185883 bytes)
gc-sys: 57.56MB (60357048 bytes)
next-gc: when heap-alloc >= 1.31GB (1402045504 bytes)
last-gc: 2020-04-09 11:38:22.480860164 +0000 UTC
gc-pause-total: 8.854325ms
gc-pause: 53084
num-gc: 26
enable-gc: true
debug-gc: false
```
```
alloc: 929.25MB (974393088 bytes)
total-alloc: 2.62GB (2809877984 bytes)
sys: 1.25GB (1341604411 bytes)
lookups: 0
mallocs: 14733085
frees: 5617388
heap-alloc: 929.25MB (974393088 bytes)
heap-sys: 1.18GB (1265664000 bytes)
heap-idle: 265.45MB (278347776 bytes)
heap-in-use: 941.58MB (987316224 bytes)
heap-released: 52.50MB (55050240 bytes)
heap-objects: 9115697
stack-in-use: 8.97MB (9404416 bytes)
stack-sys: 8.97MB (9404416 bytes)
stack-mspan-inuse: 12.73MB (13349760 bytes)
stack-mspan-sys: 12.78MB (13402112 bytes)
stack-mcache-inuse: 3.39KB (3472 bytes)
stack-mcache-sys: 16.00KB (16384 bytes)
other-sys: 2.15MB (2255667 bytes)
gc-sys: 46.90MB (49174936 bytes)
next-gc: when heap-alloc >= 1.38GB (1481919792 bytes)
last-gc: 2020-04-09 11:40:58.910320185 +0000 UTC
gc-pause-total: 8.815899ms
gc-pause: 771468
num-gc: 26
enable-gc: true
debug-gc: false
```

Those benchmarks were primarily testing the increasing number of Pods and the removal of function queues for all other watchers were based on the assumption the memory optimization will be similar to the one seen with the removal of the Pod's watcher function queues.

### CODE REVIEW

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-me-please

File: pkg/k8s/watchers/cilium_node.go [hunk 1]
```diff
@@ -18,20 +18,19 @@ import (
 	"context"
 	"sync"
 
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/fields"
+	"k8s.io/client-go/tools/cache"
```
Comment (tgraf) [inline pkg/k8s/watchers/cilium_node.go]: This moved imports around unnecessarily

File: pkg/k8s/watchers/endpoint.go [hunk 2]
```diff
@@ -15,21 +15,20 @@
...
-	"github.com/cilium/cilium/pkg/option"
-	"github.com/cilium/cilium/pkg/serializer"
-
 	v1 "k8s.io/api/core/v1"
 	"k8s.io/apimachinery/pkg/fields"
 	"k8s.io/apimachinery/pkg/util/wait"
 	"k8s.io/client-go/kubernetes"
 	"k8s.io/client-go/tools/cache"
+
+	"github.com/cilium/cilium/pkg/k8s"
```
Comment (tgraf) [inline pkg/k8s/watchers/endpoint.go]: Same here

File: pkg/k8s/watchers/namespace.go [hunk 3]
```diff
@@ -79,22 +69,13 @@ func (k *K8sWatcher) namespacesInit(k8sClient kubernetes.Interface, serNamespace
...
 	namespaceController.Run(wait.NeverStop)
 }
 
 func (k *K8sWatcher) updateK8sV1Namespace(oldNS, newNS *types.Namespace) error {
-	if oldNS == nil || newNS == nil {
-		return nil
-	}
-
-	// We only care about label updates
-	if comparator.MapStringEquals(oldNS.GetLabels(), newNS.GetLabels()) {
```
Comment (tgraf) [inline pkg/k8s/watchers/namespace.go:94]: Why is this removed?

**Review by tgraf — CHANGES_REQUESTED** [review]
Looks awesome. Do you know what processing takes a long time? I assume it's the processing of all pods that we should remove.

File: pkg/k8s/watchers/namespace.go (hunk 3 above)
Comment (aanm) [reply pkg/k8s/watchers/namespace.go:94]: This is already being compared in the `k8s.EqualV1Namespace(oldNS, newNS)`. I've added a comment in the commit.

**Comment by aanm** [comment]
> Looks awesome. Do you know what processing takes a long time? I assume it's the processing of all pods that we should remove.

For now yes. Once I remove the watch of all pods I'll check the remaining structures / watchers.

**Comment by aanm** [comment]
test-me-please

**Comment by aanm** [comment]
test-ginkgo-please

**Comment by aanm** [comment]
test-ginkgo

**Comment by aanm** [comment]
restart-ginkgo

**Review by tgraf — APPROVED** [review]

**Comment by aanm** [comment]
hit https://github.com/cilium/cilium/issues/10942

**Comment by aanm** [comment]
merging since travis CI for arm64 was also disabled upstream by https://github.com/cilium/cilium/pull/10978
